react/components/ComponentsList/getComponents.ts

```diff
@@ -0,0 +1,101 @@
+import { has, path, pathOr } from 'ramda'
+import { ComponentsRegistry } from 'render'
+import { SidebarComponent } from './typings'
+
+const isSamePage = (page: string, pages: string[]) => (treePath: string) => {
+  if (treePath.startsWith(page)) {
+    return true
+  }
+
+  const currentPageLevel = page.split('/').length
+  const sameLevelPages = pages.filter(
+    (p: string) => p.split('/').length === currentPageLevel,
+  )
+
+  return !sameLevelPages.find((p: string) => treePath.startsWith(p))
+}
+
+const getParentContainerPropsGetter = (extensions: Extensions) => (
+  parentPath: string,
+) => {
+  return pathOr<string[], string[]>(
+    [],
+    [parentPath, 'props', 'elements'],
+    extensions,
+  )
+}
+
+const getComponentNameGetterFromExtensions = (extensions: Extensions) => (
+  treePath: string,
+) => extensions[treePath].component || ''
+
+const getComponentSchemaGetter = (
+  components: ComponentsRegistry | null,
+  extensions: Extensions,
+) => (treePath: string) => {
+  const getComponentName = getComponentNameGetterFromExtensions(extensions)
+  const component = getComponentName(treePath)
+  const getSchema = path([component, 'getSchema'], { ...components })
+  return (
+    path([component, 'schema'], { ...components }) ||
+    (typeof getSchema === 'function' && getSchema({}))
+  )
+}
+
+export function getComponents(
+  extensions: Extensions,
+  components: ComponentsRegistry | null,
+  page: string,
+  pages: string[],
+) {
+  const getParentContainerProps = getParentContainerPropsGetter(extensions)
+  const getComponentSchema = getComponentSchemaGetter(components, extensions)
+
+  return Object.keys(extensions)
+    .filter(isSamePage(page, pages))
+    .filter(treePath => {
+      const schema = getComponentSchema(treePath)
+      return !!schema && !!has('title', schema)
+    })
+    .sort((treePathA, treePathB) => {
```

Comment:
:pick: I know that this function was just moved here from another file, but I was wondering if we should start shallow/deep cloning arrays before using `sort` instead of sorting elements in-place. What do you think?

Reply:
What is the benefit?

Reply:
The benefit usually resides in preserving the original array by creating a new one. But, come to think of it, there's no problem in modifying the array in that case. Never mind. 😅

Reply:
Hm, we may want to re-use it elsewhere, right? Maybe this is a fair point...